airbyte-integrations/connectors/destination-sftp-json/destination_sftp_json/spec.json

```diff
@@ -0,0 +1,64 @@
+{
+  "documentationUrl": "https://docs.airbyte.io/integrations/destinations/sftp-json",
+  "supported_destination_sync_modes": ["overwrite", "append"],
+  "supportsIncremental": true,
+  "supportsDBT": false,
+  "supportsNormalization": false,
+  "connectionSpecification": {
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "title": "Destination SFTP JSON",
+    "type": "object",
+    "required": [
+      "host",
+      "port",
+      "username",
+      "password",
+      "destination_path"
+    ],
+    "additionalProperties": false,
+    "properties": {
+      "host": {
+        "title": "Host",
+        "description": "Hostname of the SFTP server.",
+        "type": "string",
+        "order": 0
+      },
+      "port": {
+        "title": "Port",
+        "description": "Port of the SFTP server.",
+        "type": "integer",
+        "minimum": 0,
+        "maximum": 65536,
+        "default": 22,
+        "examples": ["22"],
+        "order": 1
+      },
+      "username": {
+        "title": "User",
+        "description": "Username to use to access the SFTP server.",
+        "type": "string",
+        "order": 2
+      },
+      "password": {
+        "title": "Password",
+        "description": "Password associated with the username.",
+        "type": "string",
+        "airbyte_secret": true,
+        "order": 3
+      },
+      "destination_path": {
+        "title": "Destination path",
+        "type": "string",
+        "description": "Path to the directory where json files will be written.",
+        "examples": ["/json_data"],
+        "order": 4
+      },
+      "filename": {
+        "title": "Filename",
```

Comment:
As I'm looking at this now and seeing other connectors, should this maybe be based off of the stream name instead? I might also need to change the client logic so that it can accept the filename when writing/reading/deleting, so multiple streams can be handled at once.

Reply:
Yeah I was just going to say :) the best approach is probably to mkdir for each incoming stream name in the destination. The [S3 destination](https://docs.airbyte.io/integrations/destinations/s3) uses the following directory structure which I think makes sense here too: 

```
<bucket-name>/<sorce-namespace-if-exists>/<stream-name>/<upload-date>-<upload-mills>-<partition-id>.<format-extension>
```

Reply:
Okay looking at this once more with fresh eyes - it seems like each `message` could potentially reference a different stream? If that's the case, it might be hard to manage multiple files during a single output _over SFTP_ simultaneously. I think it might be good to build all of the files up in a local temp directory, then copy them over to the SFTP server once the input is exhausted. What do you think?

Reply:
Is there a way to grab all the messages for a particular stream up front without having to iterate over all the data?

Reply:
@AetherUnbound 

Keeping a buffer in the destination connector could get tricky depending on the source of data (e.g: if we're loading 20gb of data, that could overflow the node running it) so I'd recommend that the staging happens on SFTP itself if possible through a mechanism like the following: 
1. Accumulate the data into a staging file per stream
2. Once done accumulating all records, `echo` the file to append it to its final destination (if append is used). 
If we want to be really atomic we can instead copy the target file if it already exists into a staging directory, append to that file, then `mv` the file into its final location. 


> Is there a way to grab all the messages for a particular stream up front without having to iterate over all the data?
not always -- some stream data may be interleaved. Currently there is no guarantee otherwise. (Although we should probably introduce a message that allows the source to at least signal which streams the destination could currently receive, something like `BEGIN <stream_name>` and `END <stream_name>`, this way the destination would know which buffers to hold in memory). 

Is it possible to append the data to the end of the appropriate file? Something in principle like `echo "hi" >> file.txt`. i'm not sure what the file system implications of that would be. If the append filesystem operation is expensive we could keep buffers. 


Could we append to files the same way `echo "hello world" >> file.txt` would? I guess it's probably not great for the file system to be getting that many random writes but I'm not sure what the alternative here would be

Reply:
Got it, thanks for the in-depth explanation! I'll ponder how I can refactor this to account for streams appropriately.